securedrop/secure_tempfile.py

```diff
@@ -1,93 +1,119 @@
+    """Temporary file that provides on-the-fly encryption.
 
-    Since only encrypted data is ever written to disk, using this
-    classes minimizes the chances of plaintext recovery through
-    forensic disk analysis.
+    Buffering large submissions in memory as they come in requires too
+    much memory for too long a period. By writing the file to disk as it
+    comes in using a stream cipher, we are able to minimize memory usage
+    as submissions come in, while minimizing the chances of plaintext
+    recovery through forensic disk analysis. They key used to encrypt
+    each secure temporary file is also ephemeral, and is only stored in
+    memory only for as long as needed.
 
-    Adapted from Globaleaks' GLSecureTemporaryFile: https://github.com/globaleaks/GlobaLeaks/blob/master/backend/globaleaks/security.py#L35
+    Adapted from Globaleaks' GLSecureTemporaryFile:
+    https://github.com/globaleaks/GlobaLeaks/blob/master/backend/globaleaks/security.py#L35
 
     WARNING: you can't use this like a normal file object. It supports
-    being written to exactly once, then read from exactly once.
+    being appended to however many times you wish (although content may not be
+    overwritten), and then it's contents may be read only once (although it may
+    be done in chunks) and only after it's been written to.
     """
-
     AES_key_size = 256
     AES_block_size = 128
 
     def __init__(self, store_dir):
+        """Generates an AES key and an initialization vector, and opens
+        a file in the `store_dir` directory with a
+        pseudorandomly-generated filename.
+
+        Args:
+            store_dir (str): the directory to create the secure
+                temporary file under.
+
+        Returns: self
+        """
         self.last_action = 'init'
         self.create_key()
-
         self.tmp_file_id = base64.urlsafe_b64encode(os.urandom(32)).strip('=')
-        self.filepath = os.path.join(
-            store_dir,
-            "{}.aes".format(
-                self.tmp_file_id))
+        self.filepath = os.path.join(store_dir,
+                                     '{}.aes'.format(self.tmp_file_id))
         self.file = open(self.filepath, 'w+b')
-
-        _TemporaryFileWrapper.__init__(
-            self,
-            self.file,
-            self.filepath,
-            delete=True)
+        super(SecureTemporaryFile, self).__init__(self.file, self.filepath)
 
     def create_key(self):
-        """
-        Randomly generate an AES key to encrypt the file
+        """Generates a unique, pseudorandom AES key, stored ephemerally
+        in memory as an instance attribute. Its destruction is ensured
+        by the Python garbage collector and the PAX_MEMORY_SANITIZE
+        feature of the grsecurity-patched kernel used in SecureDrop (for
+        further details consult
+        https://github.com/freedomofpress/securedrop/pull/477#issuecomment-168445450).
```

Comment:
Sorry, this is still inaccurate. You can't say "it's destruction is ensured by the Python garbage collector." It's the precise *lack* of assurance over memory allocation lifetimes due to the Python garbage collector that lead us to implement the "nightly reboot" kludge described in the link GitHub comment.